### PR TITLE
Opaque relay op codegen

### PR DESCRIPTION
**Pull request by hypercubestart**
TODO:
- [x] implement codegen for relay ops
- [x] implement add broadcasting
- [x] add tests
  - [x] RelayGlobalAvgPool2D
  - [x] RelayBatchFlatten
  - [x] RelayBiasAdd
- [x] implement assertion to check `result` with `out` array (need to add functions on NDarray)
  - [x] batchnorm
  - [x] softmax
  - [x] relu
  - [x] maxpool2d

codegen for #65

### CODE REVIEW

**Comment by hypercubestart** [comment]
@gussmith23 build is failing because CI/CD machine doesn't have "math.h"

**Comment by gussmith23** [comment]
Wait until #73 merged

**Comment by gussmith23** [comment]
Interesting about math.h...odd that it wouldn't have it. Will debug later!

File: src/language/from_relay/mod.rs [hunk 1]
```diff
@@ -706,7 +706,8 @@ fn compile_expression(
                         .downcast::<tvm::ir::relay::attrs::nn::BatchNormAttrs>()
                         .unwrap();
 
-                    assert_eq!(attrs.axis, 1);
+                    // Should be NHWC or NCHW
```
Comment (gussmith23) [inline src/language/from_relay/mod.rs:709]:
```suggestion
                    // Should be ___C or _C___ (e.g. NHWC, NCHW)
```

File: src/language/from_relay/run_relay.py [hunk 2]
```diff
@@ -32,7 +32,8 @@
     expr = tvm.parser.fromtext(relay_in)
 
 inputs = [np.load(arg_file) for arg_file in parsed.npy_arg_filepath]
-
+for i in inputs:
+    print(i.shape)
```
Comment (gussmith23) [inline src/language/from_relay/run_relay.py:36]:
```suggestion
```

**Review by gussmith23 — CHANGES_REQUESTED** [review]
Looking good, let's iterate!

File: src/codegen.rs [hunk 3]
```diff
@@ -503,7 +509,494 @@ fn codegen_recursive_helper(
...
+                    let _strides = MyAnalysis::get_shape_of_value(ids[3], expr);
+                    let _padding = MyAnalysis::get_shape_of_value(ids[4], expr);
+
+                    // TODO: currently hardcoded shape for max pool2d for resnet
+                    let old_shape = match &expr[ids[1]].data {
+                        MyAnalysisData::AccessPattern(a) => a.as_vec(),
+                        _ => panic!(),
+                    };
+
+                    let new_shape = vec![old_shape[0], 56, 56, old_shape[3]];
```
[the remark this answers is not in the document]
Comment (hypercubestart) [reply src/codegen.rs:760]: yep, to not hardcode shape, the maxpool c function needs to be extended for other shapes

File: src/codegen.rs [hunk 4]
```diff
@@ -503,7 +509,494 @@ fn codegen_recursive_helper(
...
+                        hw_map,
+                    );
+
+                    // TODO: support broadcasting
+                    let old_shape = match &expr[ids[1]].data {
+                        MyAnalysisData::AccessPattern(a) => a.as_vec(),
+                        _ => panic!(),
+                    };
+
+                    let new_shape = vec![old_shape[0], old_shape[3]];
```
[the remark this answers is not in the document]
Comment (hypercubestart) [reply src/codegen.rs:824]: dont quite follow here. global avg pool 2d averages across height/width layers, so an input dim= 4 NHWC tensor becomes a dim = 2 NC tensor

[the remark this answers is not in the document]
Comment (hypercubestart) [reply src/codegen.rs:824]: i dont think quite follow here. The input tensor is NHWC, and the output tensor is expected to be NC. Is there another way to getting the shape?

File: src/codegen.rs [hunk 5]
```diff
@@ -2447,4 +3011,1210 @@ int main() {{
...
+        ));
+        println!("{}", binary_filepath.to_string_lossy());
+
+        File::create(&main_c_filepath)
+            .unwrap()
+            .write_all(main_code.as_bytes())
+            .unwrap();
+
+        // TODO: find a better way to convert from C multidimensional array to pointer
+        // rather than removing -Werror
```
[the remark this answers is not in the document]
Comment (hypercubestart) [reply src/codegen.rs:3470]: not much of a better solution maybe, but I just added explicit casts

File: src/codegen.rs [hunk 6]
```diff
@@ -2447,4 +3011,1210 @@ int main() {{
...
+            result_output.shape().iter().product::<usize>()
+        );
+
+        println!("{}", main_code);
+
+        let main_c_filepath = std::env::temp_dir().with_file_name(format!(
+            "relay-op-softmax-test-{}.c",
+            std::time::SystemTime::now().elapsed().unwrap().as_nanos()
+        ));
+        println!("{}", main_c_filepath.to_string_lossy());
```
[the remark this answers is not in the document]
Comment (hypercubestart) [reply src/codegen.rs:3606]: i think this should be included for debugging sake...

File: src/codegen.rs [hunk 7]
```diff
@@ -2447,4 +3011,1210 @@ int main() {{
...
+{}
+{}
+{}
+{}
+
+int main() {{
+    relay_globalavgpool2d(out, x);
+
+  for (int i = 0; i < {}; i++) {{
+    assert(fabs(((float*)result)[i] - ((float*)out)[i]) < 0.001);
```
[the remark this answers is not in the document]
Comment (hypercubestart) [reply src/codegen.rs:4149]: works!

File: src/codegen.rs [hunk 8]
```diff
@@ -2447,4 +3011,1210 @@ int main() {{
...
+            "relay-op-softmax-test-{}.c",
+            std::time::SystemTime::now().elapsed().unwrap().as_nanos()
+        ));
+        println!("{}", main_c_filepath.to_string_lossy());
+
+        let binary_filepath = std::env::temp_dir().with_file_name(format!(
+            "relay-op-softmax-test-{}",
+            std::time::SystemTime::now().elapsed().unwrap().as_nanos()
+        ));
+        println!("{}", binary_filepath.to_string_lossy());
```
[the remark this answers is not in the document]
Comment (hypercubestart) [reply src/codegen.rs:3569]: i think this should be included for debugging sake...

File: c-files/relay-op-implementations.c [hunk 9]
```diff
@@ -22,6 +22,8 @@
 void batchNormInference(float *X, float *Y, int N, int H, int W, int C,
                         float *gamma, float *beta, float *mu, float *var,
                         float epsilon) {
+  //TODO: (1/sqrt(var+epsilon) * gamma) can be pre-computed
```
Comment (gussmith23) [inline c-files/relay-op-implementations.c:25]:
```suggestion
  // TODO: (1/sqrt(var+epsilon) * gamma) can be pre-computed
```

File: src/codegen.rs [hunk 10]
```diff
@@ -503,7 +504,486 @@ fn codegen_recursive_helper(
...
+                        hw_map,
+                    );
+
+                    let axis = MyAnalysis::get_usize(ids[6], expr);
+
+                    // Pre-ISCA: expect NHWC format data
+                    assert!(axis == 3, "expected NHWC format");
+                    let epsilon = match &expr[ids[7]].data {
+                        MyAnalysisData::Literal(l) => {
+                            println!("shape of batchnorm inference epsilon: {:?}", &l.shape());
```
Comment (gussmith23) [inline src/codegen.rs:569]:
```suggestion
```
I assume we don't need this anymore? Can you check for other leftover print statements?

File: src/codegen.rs [hunk 11]
```diff
@@ -2447,4 +2998,1150 @@ int main() {{
                 .expect("Could not convert stderr to UTF8")
         );
     }
+
+    #[test]
+    fn relay_op_add() {
+        // TODO: do broadcasting
```
Comment (gussmith23) [inline src/codegen.rs:3004]:
```suggestion
```
Seems like it's out of date?

File: src/codegen.rs [hunk 12]
```diff
@@ -2447,4 +2998,1150 @@ int main() {{
...
+            result.status.success(),
+            "{}",
+            std::str::from_utf8(result.stderr.as_slice())
+                .expect("Could not convert stderr to UTF8")
+        );
+    }
+
+    #[test]
+    fn relay_op_biasadd() {
+        // TODO: do broadcasting
```
Comment (gussmith23) [inline src/codegen.rs:3141]:
```suggestion
```

I think this one is also out of date?

File: src/codegen.rs [hunk 13]
```diff
@@ -2447,4 +2998,1150 @@ int main() {{
...
+
+        let code = codegen(
+            &egraph,
+            id,
+            &HashMap::default(),
+            "relay_batchnorm",
+            "",
+            &vec!["data", "bn_gamma", "bn_beta", "bn_mean", "bn_var"],
+        );
+        // TODO: check out array with result array
```
Comment (gussmith23) [inline src/codegen.rs:3342]: What does this mean? Is it still relevant?

File: src/codegen.rs [hunk 14]
```diff
@@ -2447,4 +2998,1150 @@ int main() {{
...
+
+        let mut egraph = EGraph::new(MyAnalysis {
+            name_to_shape: env.clone(),
+        });
+
+        let id = egraph.add_expr(&expr);
+
+        let result = run_relay(&value_env, &shapes_vec, relay_to_run);
+
+        println!("{}", expr);
```
Comment (gussmith23) [inline src/codegen.rs:3332]:
```suggestion
```
Remove this and other printlns

File: src/codegen.rs [hunk 15]
```diff
@@ -503,7 +504,486 @@ fn codegen_recursive_helper(
...
+                        ids[2],
+                        top_level_id,
+                        uninitialized_allocations_prefix,
+                        declarations,
+                        code,
+                        hw_map,
+                    );
+
+                    // TODO: support broadcasting
+                    // TODO: cannot assume adding 4d tensors...
```
Comment (gussmith23) [inline src/codegen.rs:899]: Are these both still relevant?

File: src/codegen.rs [hunk 16]
```diff
@@ -2447,4 +2998,1150 @@ int main() {{
...
+
+        let code = codegen(
+            &egraph,
+            id,
+            &HashMap::default(),
+            "relay_softmax",
+            "",
+            &vec!["data"],
+        );
+        // TODO: check out array with result array
```
Comment (gussmith23) [inline src/codegen.rs:3511]: Another one -- is this still relevant?

File: src/codegen.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (gussmith23) [reply src/codegen.rs:3563]: Sounds good!

**Review by gussmith23 — COMMENTED** [review]
Nearly there, just a few questions!

File: src/codegen.rs [hunk 17]
```diff
@@ -627,11 +624,10 @@ batchNormInference((float*) {X}, (float*) {Y}, {N}, {H}, {W}, {C}, (float*) {gam
...
                     );
-
-                    let axis = MyAnalysis::get_usize(ids[2], expr);
-
-                    // Pre-ISCA: resnet only does softmax over (1,1000)
-                    assert!(axis == 1, "expected NHWC format");
+                        
+                    // TODO: axis not used since
+                    // softmax c function does softmax over all values
+                    let _axis = MyAnalysis::get_usize(ids[2], expr);
```
Comment (gussmith23) [inline src/codegen.rs:630]: Hm. The input shape here is `(1, 1000)`, right? As long as the first dimension, batch, is 1, then our softmax kernel works. Can you add an assert here to assert that batch is 1?

Comment (gussmith23) [reply src/codegen.rs:630]: That is, if we have batch > 1, the kernel breaks (because it will softmax over EVERYTHING rather than splitting the kernel up), but as long as batch == 1, then we're good.

**Review by gussmith23 — COMMENTED** [review]
Can you run the formatter, too? Looks like formatting got out of whack.

Plus one last comment, then we're good!

File: src/codegen.rs (hunk 17 above)
Comment (hypercubestart) [reply src/codegen.rs:630]: is the shape guaranteed to be two dimensions though?

Comment (gussmith23) [reply src/codegen.rs:630]: Yeah, that should be a fine assumption for now. All the CNNs we're currently working with use softmax as their very last layer, and the input/output shape is `(batch, num_classes)`. You could also assert that the number of dimensions is 2, if you want to be extra safe. That should be the case for all of our current networks.

Comment (gussmith23) [reply src/codegen.rs:630]: FWIW, my general strategy is that it's okay if we put in asserts that OVER-constrain. Then, at the very worst, you get a false-positive assertion failure, and you can just refine the condition in the assertion. In your current code, I don't think there's anything to stop us from generating C code that just straight up breaks, if the batch size is greater than 1 in the softmax, or if the `axis` param from relay is something out of the norm. Then, the bug would come in the form of an error from the C code and would be harder to track down!

**Review by gussmith23 — APPROVED** [review]
